# 2 QB Data Wrangling

   ## 2.1 Contents<a id='2.1_Contents'></a>
* [2 Data wrangling](#2_Data_wrangling)
  * [2.1 Contents](#2.1_Contents)
  * [2.2 Introduction](#2.2_Introduction)
    * [2.2.1 Recap Of Data Science Problem](#2.2.1_Recap_Of_Data_Science_Problem)
    * [2.2.2 Introduction To Notebook](#2.2.2_Introduction_To_Notebook)
  * [2.3 Imports](#2.3_Imports)
  * [2.4 Objectives](#2.4_Objectives)
  * [2.5 Load The QB Data](#2.5_Load_The_QB_Data)
    * [2.5.1 Clean the QB Data](#2.5.1_Clean_the_QB_Data)
    * [2.5.2 Adding Rows](#2.5.2_Adding_Rows)
    * [2.5.3 Rearrange Columns](#2.5.3_Rearrange_Columns)
  * [2.6 Explore the Data](#2.6_Explore_the_Data)
    * [2.6.1 Create Values for the NaN Columns](#2.6.1_Create_Values_for_the_NaN_Columns)
      * [2.6.1.1 Create Values for td per cmp](#2.6.1.1_Create_Values_for_td_per_cmp)
      * [2.6.1.2 Create Values for ypc](#2.6.1.2_Create_Values_for_ypc)
      * [2.6.1.3 Create Values for QB Rating](#2.6.1.3_Create_Values_for_QB_Rating)
    * [2.6.2 Looking at NaN Values](#2.6.2_Looking_at_NaN_Values)
  * [2.7 Explore the Columns](#2.7_Explore_the_Columns) 
      * [2.7.1 QB Column](#2.7.1_QB_Column)
      * [2.7.2 CMP Column](#2.7.2_CMP_Column)
      * [2.7.3 ATT Column](#2.7.3_ATT_Column)
      * [2.7.4 Comp % Column](#2.7.4_Comp_%_Column)
      * [2.7.5 YDS Column](#2.7.5_yds_Column)
      * [2.7.6 TD Column](#2.7.6_td_Column)
      * [2.7.7 INT Column](#2.7.7_int_Column)
      * [2.7.8 Rate Column](#2.7.8_Rate_Column)
      * [2.7.9 Long Column](#2.7.9_long_Column)
      * [2.7.10 Sack Column](#2.7.10_Sack_Column)
      * [2.7.11 Game Points Column](#2.7.11_Game_Points_Column)
      * [2.7.12 YPA Column](#2.7.12_YPA_Column)
      * [2.7.13 TD_per CMP Column](#2.7.13_TD_per_CMP_Column)
      * [2.7.14 TD_per ATT Column](#2.7.14_TD_per_ATT_Column)
      * [2.7.15 loss_YDS Column](#2.7.15_Loss_YDS_Column)
      * [2.7.16 YPC Column](#2.7.16_YPC_Column)
      * [2.7.17 Home_Away Column](#2.7.17_Home_Away_Column)
      * [2.7.18 Year Column](#2.7.18_Year_Column)
  * [2.8 Taking a Look at the Data as a Whole](#2.8_Taking_a_Look_at_the_Data_as_a_Whole) 
  * [2.9 Summary](#2.9_Summary) 


## 2.2 Introduction<a id='2.2_Introduction'></a>

This section is about organizing the QB data and making sure it's well defined. Doing this will help imensely in the future to get what we need out of the data. Some data cleaning will be done at this stage, but we will get into cleaning more as we understand what we need in later parts of the data exploration.

### 2.2.1 Recap Of Data Science Problem<a id='2.2.1_Recap_Of_Data_Science_Problem'></a>

The purpose of this data science project is to evaluate Quarterbacks in the most efficient manner. A NFL team needs help in assessing the Quarterback position with more information than the usual Passer Rating. The QB position is arguably the most important position on the field as they touch the ball on nearly every play. Their decisions can have a huge effect on the outcome of the game. This model will help teams evaluate the QBs on whom to offer a contract, draft, bench or start.

### 2.2.2 Introduction To Notebook<a id='2.2.2_Introduction_To_Notebook'></a>

In Jupyter notebooks you can work in sequence. Here you can add, edit and rearrange cells around without needing to cross out code or wording. Every new cell is an oppurtunity to express a idea.

# 2.3 Imports<a id='2.3_Imports'></a>

Here we will use os, pandas, mathplotlib, seaborn and numpy to access and manipulate the QB data. Placing your imports all together at the start of your notebook means you only need to check one place for your notebook's dependencies.

In [4]:
import os
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

## 2.4 Objectives<a id='2.4_Objectives'></a>

There are four objectives:

1. Do we think we have the correct data for our desired question?

2. Have we identified the required target value?

3. Do we have features that are potentially useful?

4. Are there fundamental issues with the data?

## 2.5 Load The QB Data<a id='2.5_Load_The_QB_Data'></a>

In [5]:
# this is the supplied QB CSV data file 
df= pd.read_csv('QBStats_all.csv')

Now we need to see the first five lines to see what the data looks like just t make sure everythign looks ok! I noticed that there was NaN value for lg when I called the .info(). We will change that soon.

In [6]:
#call .head() to see the first few rows 
df.head()

qb  att  cmp    yds  ypa  td int   lg  sack  loss  \
0  Boomer EsiasonB. Esiason   38   25  237.0  6.2   0   0   20   2.0  11.0   
1   Jim HarbaughJ. Harbaugh   25   16  196.0  7.8   2   1  35t   0.0   0.0   
2      Paul JustinP. Justin    8    5   53.0  6.6   0   0   30   1.0  11.0   
3      Jeff GeorgeJ. George   35   16  215.0  6.1   0   0   55   7.0  53.0   
4   Kerry CollinsK. Collins   31   17  198.0  6.4   2   0   30   4.0  12.0   

   rate  game_points home_away  year  
0  82.9           13      away  1996  
1  98.1           20      home  1996  
2  81.8           20      home  1996  
3  65.8            6      away  1996  
4  95.9           29      home  1996

### 2.5.1 Clean the QB Data<a id='2.5.1_Clean_the_QB_Data'></a>

First lets rename 'lg' to 'long' and 'loss' to 'loss_yds' to be clear on what the column represents. 

In [7]:
#rename columns
dict = {'lg': 'long',
       'loss': 'loss_yds'}
 
df.rename(columns=dict,
          inplace=True)

Here we can see that there is a 't' in the lg column. This represents that the QB had their longest pass of 35 yard pass that resulted into a td. we need to get rid of all the 't' and turn lg into an integer. Also lets change the NaN values of long to 0 since a QB that did not complete a pass was filled in as NaN and needs to be numeric.

In [8]:
#remove all t from long
df['long'] = df['long'].str.replace('t', '')
#replace all NaN values with 0
df['long'] = df['long'].fillna(0)
#long into a int
df['long']=df.long.astype(int)

### 2.5.2 Adding Rows<a id='2.5.2_Adding_Rows'></a>

We also need to create a few more columns to help us grade a QB on their preformance. Here we will make a column for:
1. Completion % 
2. Yards per completion
3. Touchdown passes per attempts %
4. Touchdown passes per completion %

In [9]:
#create completion %
df['comp %'] = df['cmp'] / df['att'] 
df['comp %'] = round(df['comp %']*100,1)


#create yards per completion
df['ypc'] = df['yds'] / df['cmp']
df['ypc'] =round(df['ypc'], 1)


#Touchdown Passes per attempts % = Touchdown Passes / Pass Attempts
df['td_per_att'] = df['td'] / df['att']
df['td_per_att'] =round(df['td_per_att'], 3)


#Touchdown Passes per completion % = Touchdown Passes / Pass completions
df['td_per_cmp'] = df['td'] / df['cmp']
df['td_per_cmp'] =round(df['td_per_cmp'], 3)

### 2.5.3 Rearrange Columns<a id='2.5.3_Rearrange_Columns'></a>

Lets rearrange the columns to make it look nice and put relavant columns next to each other.

In [10]:
#reargange columns
df = df[['qb', 'cmp', 'att', 'comp %', 'yds', 'td', 'int', 'rate',  'long', 'sack', 'game_points', 
          'ypa','ypc', 'td_per_cmp', 'td_per_att', 'loss_yds', 'home_away', 'year']]

Lets take a look at the four new columns, the rearganed columns and to see if the 't' was removed from the long column.

In [11]:
df.head()

qb  cmp  att  comp %    yds  td int  rate  long  \
0  Boomer EsiasonB. Esiason   25   38    65.8  237.0   0   0  82.9    20   
1   Jim HarbaughJ. Harbaugh   16   25    64.0  196.0   2   1  98.1    35   
2      Paul JustinP. Justin    5    8    62.5   53.0   0   0  81.8    30   
3      Jeff GeorgeJ. George   16   35    45.7  215.0   0   0  65.8    55   
4   Kerry CollinsK. Collins   17   31    54.8  198.0   2   0  95.9    30   

   sack  game_points  ypa   ypc  td_per_cmp  td_per_att  loss_yds home_away  \
0   2.0           13  6.2   9.5       0.000       0.000      11.0      away   
1   0.0           20  7.8  12.2       0.125       0.080       0.0      home   
2   1.0           20  6.6  10.6       0.000       0.000      11.0      home   
3   7.0            6  6.1  13.4       0.000       0.000      53.0      away   
4   4.0           29  6.4  11.6       0.118       0.065      12.0      home   

   year  
0  1996  
1  1996  
2  1996  
3  1996  
4  1996

Looks good!

## 2.6 Explore the Data<a id='2.6_Explore_the_Data'></a>

Count the number of missing values in each column and sort them.

In [12]:
missing = pd.concat([df.isnull().sum(), 100 * df.isnull().mean()], axis=1)
missing.columns=["count", '%']
missing.sort_values(by='count', ascending=False)

count         %
td_per_cmp     719  5.451926
ypc            719  5.451926
comp %          97  0.735517
td_per_att      97  0.735517
sack            17  0.128905
loss_yds        17  0.128905
rate            17  0.128905
game_points      0  0.000000
home_away        0  0.000000
ypa              0  0.000000
qb               0  0.000000
cmp              0  0.000000
long             0  0.000000
int              0  0.000000
td               0  0.000000
yds              0  0.000000
att              0  0.000000
year             0  0.000000

td_per_comp and ypc, (yards per catch) have the most missing values at 5.45%. Comp % and td_per_att are missing .735% of their respective data. Sack, loss_yds, and rate are missing an equal amount as well at .128%. We need to investigate the missing data and decide what to enter for this missing data. Sucn as zero, the mean or something else.

In [13]:
df[df.isnull().any(axis=1)].head(20)

qb  cmp  att  comp %  yds  td int  rate  long  \
19      Kordell StewartK. Stewart    0    2     0.0  0.0   0   0  39.6     0   
28            Elvis GrbacE. Grbac    0    2     0.0  0.0   0   0  39.6     0   
32       Bill MusgraveB. Musgrave    0    1     0.0  0.0   0   0  39.6     0   
55                Tom TupaT. Tupa    0    1     0.0  0.0   0   0  39.6     0   
66     Sean SalisburyS. Salisbury    0    1     0.0  0.0   0   1   0.0     0   
75   Lawrence PhillipsL. Phillips    0    0     NaN  0.0   0   0   0.0     0   
85      Craig HentrichC. Hentrich    0    1     0.0  0.0   0   0  39.6     0   
89        David MeggettD. Meggett    0    1     0.0  0.0   0   0  39.6     0   
124        Brad JohnsonB. Johnson    0    1     0.0  0.0   0   0  39.6     0   
136   Sherman WilliamsS. Williams    0    1     0.0  0.0   0   0  39.6     0   
144    Harvey WilliamsH. Williams    0    1     0.0  0.0   0   0  39.6     0   
151          Paul JustinP. Justin    0    1     0.0  0.0   0   0  39.6     0   
156         Aaron CraverA. Craver    0    1     0.0  0.0   0   0  39.6     0   
185         Bobby HoyingB. Hoying    0    0     NaN  0.0   0   0   0.0     0   
230           Glenn FoleyG. Foley    0    1     0.0  0.0   0   0  39.6     0   
236     Brian MitchellB. Mitchell    0    1     0.0  0.0   0   0  39.6     0   
249        Browning NagleB. Nagle    0    0     NaN  0.0   0   0   0.0     0   
271     Kordell StewartK. Stewart    0    1     0.0  0.0   0   0  39.6     0   
295           Terry KirbyT. Kirby    0    1     0.0  0.0   0   0  39.6     0   
305      Bill MusgraveB. Musgrave    0    1     0.0  0.0   0   1   0.0     0   

     sack  game_points  ypa  ypc  td_per_cmp  td_per_att  loss_yds home_away  \
19    0.0            9  0.0  NaN         NaN         0.0       0.0      away   
28    0.0           27  0.0  NaN         NaN         0.0       0.0      home   
32    0.0           31  0.0  NaN         NaN         0.0       0.0      home   
55    0.0           10  0.0  NaN         NaN         0.0       0.0      away   
66    0.0           27  0.0  NaN         NaN         0.0       0.0      home   
75    1.0            0  0.0  NaN         NaN         NaN      12.0      away   
85    0.0           39  0.0  NaN         NaN         0.0       0.0      home   
89    0.0           31  0.0  NaN         NaN         0.0       0.0      home   
124   0.0           30  0.0  NaN         NaN         0.0       0.0      home   
136   0.0            7  0.0  NaN         NaN         0.0       0.0      away   
144   0.0           34  0.0  NaN         NaN         0.0       0.0      home   
151   0.0           10  0.0  NaN         NaN         0.0       0.0      home   
156   0.0           14  0.0  NaN         NaN         0.0       0.0      away   
185   1.0           19  0.0  NaN         NaN         NaN      10.0      home   
230   0.0           17  0.0  NaN         NaN         0.0       0.0      away   
236   0.0           27  0.0  NaN         NaN         0.0       0.0      away   
249   1.0           28  0.0  NaN         NaN         NaN       2.0      away   
271   0.0           13  0.0  NaN         NaN         0.0       0.0      away   
295   0.0           10  0.0  NaN         NaN         0.0       0.0      away   
305   0.0           34  0.0  NaN         NaN         0.0       0.0      home   

     year  
19   1996  
28   1996  
32   1996  
55   1996  
66   1996  
75   1996  
85   1996  
89   1996  
124  1996  
136  1996  
144  1996  
151  1996  
156  1996  
185  1996  
230  1996  
236  1996  
249  1996  
271  1996  
295  1996  
305  1996

Here we can see that the 97 comp % and td_per att are NaN values because they did not throw a pass but were credited as having a sack, (where they were going to attempt a pass but never even made it). These 97 data points dont really help us to determine a Qb rating system. We should get rid of these values since they are outliers during the data clean phase.

The loss_yds and sack are correlated with 17 NaN values since the QB is accredited with loss yards for every sack that occurs. A sack is when a QB drops back to pass but is tackle behind the line of scrimmage. We will fill them in with 0s since they still have other relavant data.

The rate, td_per_cmp and ypc columns do not show any NaN values.

### 2.6.1 Create Values for the NaN Columns<a id='2.6.1_Create_Values_for_the_NaN_Columns'></a>

1. Lets take a look at the Interception 'int' column first. AN interception is when a QB throws the ball to the other team.

In [14]:
df['int'].value_counts()

0     6384
1     4089
2     1870
3      624
4      166
5       32
--      17
6        5
7        1
Name: int, dtype: int64

Here we will enter 0 for the '--'. We can assume they meant that the QB did not have an interception for that particular game.

In [15]:
df['int'] = df['int'].replace('--', 0)

We need to turn the 'int' into a integer as well.

In [16]:
#int into float
df['int']=df.int.astype(int)

### 2.6.1.1 Create Values for td per cmp<a id='2.6.1.1_Create_Values_for_td_per_cmp'></a>

2. Now lets look at the touchdown per completion/'td_per_cmp' column.

In [17]:
df['td_per_cmp'].head(20)

0     0.000
1     0.125
2     0.000
3     0.000
4     0.118
5     0.043
6     0.000
7     0.050
8     0.000
9     0.062
10    0.000
11    0.095
12    0.062
13    0.118
14    0.000
15    0.100
16    0.000
17    0.000
18    0.000
19      NaN
Name: td_per_cmp, dtype: float64

In [18]:
#replace all td per cmp NaN values with 0
df['td_per_cmp'] = df['td_per_cmp'].fillna(0)

The 'td_per_cmp' column looks good!

### 2.6.1.2 Create Values for ypc<a id='2.6.1.2_Create_Values_for_ypc'></a>

3. Lets look at the yards per completion/'ypc' column

In [19]:
df['ypc'].head(20)

0      9.5
1     12.2
2     10.6
3     13.4
4     11.6
5      9.8
6     11.8
7     13.0
8     -8.0
9      9.8
10    11.4
11     9.1
12    13.5
13    11.3
14    13.4
15    13.4
16     9.9
17     9.2
18     8.0
19     NaN
Name: ypc, dtype: float64

In [20]:
#replace all ypc NaN values with 0
df['ypc'] = df['ypc'].fillna(0)

Looks like this column is good! There are no NaN values.

### 2.6.1.3 Create Values for QB Rating<a id='2.6.1.3_Create_Values_for_QB_Rating'></a>

4. Lets look at the  QB rating/'rate' data. A standard QB rating is based on four metrics: completion percentage, yards per passing attempt, touchdown percentage, and interception percentage.

In [21]:
df['rate'].value_counts(20)

39.6     0.043960
0.0      0.016779
118.8    0.015033
158.3    0.009870
56.2     0.006302
           ...   
36.1     0.000076
0.4      0.000076
24.2     0.000076
17.5     0.000076
126.2    0.000076
Name: rate, Length: 1420, dtype: float64

Looks like there are no NaN values. We can fill in NaN with 0 just to be sure. Here we have lot of different rate values which is good for our data. This could give us a wide range to determine what makes a QB good or bad. 

In [22]:
#replace all rate NaN values with 0
df['rate'] = df['rate'].fillna(0)

### 2.6.2 Looking at NaN Values<a id='2.6.2_Looking_at_NaN_Values'></a>

In [23]:
missing = pd.concat([df.isnull().sum(), 100 * df.isnull().mean()], axis=1)
missing.columns=["count", '%']
missing.sort_values(by='count', ascending=False)

count         %
comp %          97  0.735517
td_per_att      97  0.735517
sack            17  0.128905
loss_yds        17  0.128905
game_points      0  0.000000
home_away        0  0.000000
td_per_cmp       0  0.000000
ypc              0  0.000000
ypa              0  0.000000
qb               0  0.000000
cmp              0  0.000000
long             0  0.000000
rate             0  0.000000
int              0  0.000000
td               0  0.000000
yds              0  0.000000
att              0  0.000000
year             0  0.000000

During the data cleaning phase we will determine what to do with the comp %, td_per_att, sack and loss_yds columns.

## 2.7 Explore the Columns<a id='2.7_Explore_the_Columns'></a>

Here we will find out what the values, range, mean, counts, unique values, median, mode and standard deviation is for each column.

### 2.7.1 QB Column<a id='2.7.1_QB_Column'></a>

In [24]:
df['qb'].dtypes

dtype('O')

The QB column is an object.

In [25]:
df['qb'].unique()

array(['Boomer EsiasonB.\xa0Esiason', 'Jim HarbaughJ.\xa0Harbaugh',
       'Paul JustinP.\xa0Justin', 'Jeff GeorgeJ.\xa0George',
       'Kerry CollinsK.\xa0Collins', 'Jeff BlakeJ.\xa0Blake',
       'Steve WalshS.\xa0Walsh', 'Scott MitchellS.\xa0Mitchell',
       'Mark RoyalsM.\xa0Royals', 'Brad JohnsonB.\xa0Johnson',
       'Warren MoonW.\xa0Moon', 'Steve BonoS.\xa0Bono',
       'Chris ChandlerC.\xa0Chandler', 'Billy Joe HobertB.\xa0Hobert',
       'Vinny TestaverdeV.\xa0Testaverde', 'Rodney PeeteR.\xa0Peete',
       'Gus FrerotteG.\xa0Frerotte', 'Jim MillerJ.\xa0Miller',
       'Mike TomczakM.\xa0Tomczak', 'Kordell StewartK.\xa0Stewart',
       'Mark BrunellM.\xa0Brunell', 'Brett FavreB.\xa0Favre',
       'Jim McMahonJ.\xa0McMahon', 'Trent DilferT.\xa0Dilfer',
       'Drew BledsoeD.\xa0Bledsoe', 'Dan MarinoD.\xa0Marino',
       'Jim EverettJ.\xa0Everett', 'Steve YoungS.\xa0Young',
       'Elvis GrbacE.\xa0Grbac', "Neil O'DonnellN.\xa0O'Donnell",
       'Frank ReichF.\xa0Reich', 'John 

There are 13,187 unique QB values.

In [26]:
df['qb'].value_counts().head()

Peyton ManningP. Manning    265
Tom BradyT. Brady           237
Brett FavreB. Favre         235
Drew BreesD. Brees          232
Eli ManningE. Manning       199
Name: qb, dtype: int64

Peyton Manning played in the most games from 1996-2016.

In [27]:
qb_rating_means = df.groupby(by='qb')['rate'].mean().sort_values(ascending=False)
qb_rating_means.head()

qb
Josh MillerJ. Miller          158.3
Tyrone WheatleyT. Wheatley    158.3
Brandon BanksB. Banks         158.3
Ken DilgerK. Dilger           158.3
Doug BaldwinD. Baldwin        158.3
Name: rate, dtype: float64

In [140]:
df['qb'].isnull().values.any()

False

Looks like a few people have a perfect QB rating of 158.3.

### 2.7.2 CMP Column<a id='2.7.2_CMP_Column'></a>

In [28]:
df['cmp'].dtypes

dtype('int64')

It is an integer. Which it should be!

In [29]:
df['cmp'].max()

58

In [30]:
df['cmp'].min()

-6

Here we will have to get rid of the -6 value because it is impossible to throw -6 completions. Lets see if there are any other rows with a value below 0.

In [31]:
df['cmp'].max()-df['cmp'].min()

64

In [32]:
df[df['cmp'].isin([-6,-5,-4,-3,-2,-1])]

qb  cmp  att  comp %   yds  td  int  rate  \
8597        Chris GreisenC. Greisen   -1    1  -100.0 -1.00   0    0   0.0   
8602  LaBrandon ToefieldL. Toefield   -1    3   -33.3 -0.33   0    0   0.0   
8647   Brooks BollingerB. Bollinger   -6    3  -200.0 -2.00   0    0   0.0   

      long  sack  game_points  ypa  ypc  td_per_cmp  td_per_att  loss_yds  \
8597     0   NaN           24 -1.0  1.0        -0.0         0.0       NaN   
8602     0   NaN            6  1.0  0.3        -0.0         0.0       NaN   
8647     0   NaN           34  0.0  0.3        -0.0         0.0       NaN   

     home_away  year  
8597      away  2009  
8602      home  2009  
8647      away  2009

We will have to get rid of these three rows as they are incorrect.

In [33]:
df.drop([8597,8602,8647], axis=0, inplace=True)

In [34]:
df['cmp'].min()

0

It worked!

In [35]:
df['cmp'].mean() 

16.126052332195677

In [36]:
df['cmp'].median()

17.0

In [37]:
df['cmp'].mode()

0    0
dtype: int64

In [38]:
df['cmp'].std()

8.771335383519725

In [39]:
df['cmp'].unique()

array([25, 16,  5, 17, 23, 13, 20,  1, 21, 19, 12,  9,  3,  0, 18,  7,  4,
       24, 11, 22, 14, 27, 15, 26,  2, 10,  6,  8, 37, 29, 28, 32, 31, 33,
       30, 35, 34, 38, 42, 36, 39, 40, 43, 46, 58, 45, 41], dtype=int64)

In [40]:
df['cmp'].value_counts().head()

0     719
18    688
17    676
1     666
21    661
Name: cmp, dtype: int64

In [139]:
df['cmp'].isnull().values.any()

False

The mean and median are very close. The most frequent number of completions is 0. The standard deviation is pretty spread out. We have a unique range of values from 0-58. 

### 2.7.3 ATT Column<a id='2.7.3_ATT_Column'></a>

In [41]:
df['att'].dtypes

dtype('int64')

In [42]:
df['att'].max()

69

In [43]:
df['att'].min()

0

In [44]:
df['att'].mean() 

26.87607129313614

In [45]:
df['att'].median() 

29.0

In [46]:
df['att'].mode() 

0    1
dtype: int64

In [47]:
df['att'].value_counts().head()

1     1003
30     495
33     479
35     474
34     472
Name: att, dtype: int64

In [48]:
df['att'].std() 

13.580485034162818

In [49]:
df['att'].unique()

array([38, 25,  8, 35, 31, 40, 41,  1, 23, 14, 37, 29, 26, 33, 34, 17,  4,
        2, 27, 30, 22, 13, 12, 24, 21, 32, 20, 46, 19,  3, 28, 11,  6,  0,
       18, 44, 39, 36,  5,  7, 42, 54, 58, 16, 45, 10, 43, 48, 50, 15, 61,
       52, 51,  9, 49, 59, 47, 60, 53, 56, 57, 63, 55, 69, 62, 68, 64, 67,
       65], dtype=int64)

In [138]:
df['att'].isnull().values.any()

False

The mean and median are relatively close. One attempt is the most out of the value counts. The standard deviation is pretty spread out as well. We have a unique range of values from 0-69. 

### 2.7.4 Comp % Column<a id='2.7.4_Comp_%_Column'></a>

In [50]:
df['comp %'].describe()

count    13088.000000
mean        58.541626
std         20.815547
min          0.000000
25%         51.400000
50%         60.000000
75%         67.600000
max        500.000000
Name: comp %, dtype: float64

In [51]:
#replace all comp % NaN values with 0
df['comp %'] = df['comp %'].fillna(0)

In [52]:
df['comp %'].median() 

60.0

In [53]:
df['comp %'].value_counts().head(20)

0.0      719
50.0     636
100.0    560
66.7     543
60.0     372
57.1     217
75.0     194
62.5     169
55.6     167
63.6     142
71.4     140
58.3     128
61.5     118
33.3     113
54.5     109
40.0     100
70.0      98
58.8      97
61.1      91
64.3      83
Name: comp %, dtype: int64

In [137]:
df['comp %'].isnull().values.any()

False

Here we can see that we have a max value of 500% completions, (which is impossible in the game of football) we will clean this up later. The mean is 58%, so having a higher rate than this means they are doing a good job. A high std of 20.81.

### 2.7.5 YDS Column<a id='2.7.5_yds_Column'></a>

In [54]:
df['yds'].describe()

count    13185.000000
mean       186.177722
std        105.604116
min        -11.000000
25%        115.000000
50%        197.000000
75%        260.000000
max        527.000000
Name: yds, dtype: float64

In [55]:
df['yds'].median()

197.0

In [56]:
df['yds'].value_counts().head(20)

0.0      736
196.0     73
209.0     69
236.0     69
226.0     67
231.0     66
202.0     66
220.0     65
199.0     65
210.0     63
232.0     63
237.0     62
221.0     61
198.0     60
230.0     60
185.0     60
227.0     59
224.0     59
252.0     59
235.0     59
Name: yds, dtype: int64

In [136]:
df['yds'].isnull().values.any()

False

A mean of 186 yards per game. A typical very good game is 300 yards per game. The yds column is a float64 whic is good.

### 2.7.6 TD Column<a id='2.7.6_td_Column'></a>

In [57]:
df['td'].describe()

count    13185.000000
mean         1.118392
std          1.119667
min          0.000000
25%          0.000000
50%          1.000000
75%          2.000000
max          7.000000
Name: td, dtype: float64

In [58]:
df['td'].unique()

array([0, 2, 1, 4, 3, 5, 6, 7], dtype=int64)

In [59]:
df['td'].median()

1.0

In [60]:
df['td'].value_counts().head(20)

0    4793
1    4153
2    2662
3    1146
4     344
5      70
6      14
7       3
Name: td, dtype: int64

In [135]:
df['td'].isnull().values.any()

False

For the TD column we  have a mean of 1.11, 0-7 TDs were thrown in a game. Most games there waS 0-3 TDs thrown in a game.

### 2.7.7 INT Column<a id='2.7.7_int_Column'></a>

In [61]:
df['int'].describe()

count    13185.000000
mean         0.801062
std          0.970170
min          0.000000
25%          0.000000
50%          1.000000
75%          1.000000
max          7.000000
Name: int, dtype: float64

In [62]:
df['int'].unique()

array([0, 1, 2, 4, 3, 5, 6, 7])

In [63]:
df['int'].median()

1.0

In [64]:
df['int'].value_counts().head(20)

0    6398
1    4089
2    1870
3     624
4     166
5      32
6       5
7       1
Name: int, dtype: int64

In [134]:
df['int'].isnull().values.any()

False

Typically there was .8 interceptions thrown in a game and median of 1. most were between 0-3 interceptions thrown in game. one QB threw 7, yikes!

### 2.7.8 Rate Column<a id='2.7.8_Rate_Column'></a>

In [65]:
df['rate'].describe()

count    13185.000000
mean        80.157368
std         32.157365
min          0.000000
25%         58.500000
50%         80.900000
75%        102.000000
max        158.300000
Name: rate, dtype: float64

In [66]:
df['rate'].unique()

array([ 82.9,  98.1,  81.8, ..., 149.8, 150.8, 126.2])

In [67]:
df['rate'].median()

80.9

In [68]:
df['rate'].value_counts().head(20)

39.6     579
0.0      235
118.8    198
158.3    130
56.2      83
79.2      79
95.8      51
91.7      47
87.5      41
42.4      40
104.2     37
68.8      35
97.9      34
108.3     34
83.3      33
77.1      32
100.0     32
85.4      31
72.9      30
89.6      30
Name: rate, dtype: int64

In [133]:
df['rate'].isnull().values.any()

False

Here we have an average of 80.15 and standard deviation of 32.15. The different value counts more spread out than the other columns so far, but most had a value of 39.6, 0.0, 118.8 and 158.3.

### 2.7.9 Long Column<a id='2.7.9_long_Column'></a>

In [69]:
df['long'].describe()

count    13185.000000
mean        33.315738
std         18.320263
min        -11.000000
25%         22.000000
50%         31.000000
75%         44.000000
max         99.000000
Name: long, dtype: float64

In [70]:
df['long'].unique()

array([ 20,  35,  30,  55,  46,  32,  29,  -8,  31,  28,  25,  34,  33,
        48,  54,  21,  17,  10,   0,  38,  51,  24,  16,  52,  15,  50,
        39,  60,  27,   7,  37,  47,  63,  53,  71,  45,   8,  23,  12,
        22,  18,  43,  13,  62,   5,  78,  74,  11,  40,   3,  42,  80,
        14,  44,  19,  58,  61,  41,  67,  49,  64,  59,   4,  69,  26,
        36,  70,   9,  56,  65,  84,   6,  73,  82,  57,  95,  83,  66,
        77,  86,  76,  72,  89,  79,  68,   2,  75,   1,  92,  98,  81,
        -1,  87,  90,  88,  85,  -4,  94,  91,  -2, -11,  -3,  99,  -7,
        96,  93,  -6])

In [71]:
df['long'].median()

31.0

In [72]:
df['long'].value_counts().head(20)

0     752
24    387
27    368
30    366
25    358
26    348
23    336
28    333
22    332
29    331
31    315
35    308
20    300
21    296
36    287
33    282
34    273
37    265
32    262
40    243
Name: long, dtype: int64

In [132]:
df['long'].isnull().values.any()

False

The average long pass per game was 33.31 and a median of 31. The value counts are pretty evenly spread out besides the 0 value of 752.

### 2.7.10 Sack Column<a id='2.7.10_Sack_Column'></a>

In [73]:
df['sack'].describe()

count    13171.000000
mean         1.863108
std          1.701091
min          0.000000
25%          0.000000
50%          2.000000
75%          3.000000
max         12.000000
Name: sack, dtype: float64

In [74]:
df['sack'].unique()

array([ 2.,  0.,  1.,  7.,  4.,  3.,  8.,  5.,  6., 10.,  9., 12., nan,
       11.])

In [75]:
#replace all sack NaN values with 0
df['sack'] = df['sack'].fillna(0)

In [76]:
df['sack'].median()

2.0

In [77]:
df['sack'].value_counts().head(20)

0.0     3317
1.0     3156
2.0     2714
3.0     1846
4.0     1111
5.0      582
6.0      263
7.0      122
8.0       47
9.0       18
10.0       7
12.0       1
11.0       1
Name: sack, dtype: int64

In [131]:
df['sack'].isnull().values.any()

False

Here we have a mean of 1.83 sacks per game. We had a few NaN values so we changed it to 0 since there were no sacks that game. Most of the sacks per game was 0-4. There was a median of 2 per game.

### 2.7.11 Game Points Column<a id='2.7.11_Game_Points_Column'></a>

In [79]:
df['game_points'].describe()

count    13185.000000
mean        21.379446
std         10.663774
min          0.000000
25%         13.000000
50%         21.000000
75%         28.000000
max         62.000000
Name: game_points, dtype: float64

In [85]:
df['game_points'].unique()

array([13, 20,  6, 29, 16, 26, 17, 19, 14,  9, 24, 34,  3, 10, 11, 27, 31,
        7, 23, 22, 21, 30,  0, 38, 39, 15, 36, 42, 25, 35, 28, 40, 33, 18,
       12, 37, 46, 45, 32,  8, 59, 41, 55, 44, 51, 48, 50, 43,  5, 49, 47,
       57, 54, 52, 56, 58, 62,  2, 53], dtype=int64)

In [82]:
df['game_points'].median()

21.0

In [83]:
df['game_points'].value_counts().head(20)

17    873
20    869
24    821
10    767
13    720
27    661
31    596
14    549
23    544
7     519
21    494
16    426
34    415
28    379
3     360
30    329
6     319
38    303
19    291
26    268
Name: game_points, dtype: int64

In [130]:
df['game_points'].isnull().values.any()

False

Very interesting category for every NFL game from 1996-2016 the median and mean game points scored for each team was close at 21.0 and 21.37. There was a wide range of scores with 17 being the most.There was 59 different score values in the 20 years. The most points scored was 62.

### 2.7.12 YPA Column<a id='2.7.12_YPA_Column'></a>

In [87]:
df['ypa'].describe()

count    13185.000000
mean         6.884475
std          4.225494
min        -11.000000
25%          5.200000
50%          6.700000
75%          8.100000
max         81.000000
Name: ypa, dtype: float64

In [86]:
df['ypa'].unique()

array([  6.2,   7.8,   6.6,   6.1,   6.4,   5.7,   4.4,   6.3,  -8. ,
         6.8,   4.1,   5.2,   7.4,   7.7,   7.9,   4.8,   4.9,   6. ,
         0. ,   9.1,  24. ,   5.8,   8. ,   4.3,   6.9,   3.8,   5.5,
         5.1,   7. ,   7.6,  33. ,  47. ,   8.1,  10.3,  12.1,  10.6,
         6.5,   4.6,   4.5,   7.1,   3. ,   6.7,   7.5,   7.2,   2.9,
         8.4,   2.2,  11. ,   3.2,   1.8,   5. ,   5.4,  18. ,   9.5,
        11.2,   4.2,   3.5,   8.7,   5.9,   8.6,   5.3,   3.7,   7.3,
         8.5,   3.9,   9.4,   5.6,   9.9,  11.1,   9. ,   8.9,  10.2,
         9.3,  10.9,  11.7,   9.2,  16. ,   1.7,   8.8,  10.8,   4.7,
         4. ,   2.4,   8.2,   8.3,   9.7,  12. ,   3.4,  12.5,  13.2,
         9.6,  30. ,  -1.3,  13.8,   2. ,  12.6,  11.3,   3.3,  15.5,
        10.5,   3.6,   2.8,  10. ,  11.8,   2.7,   2.6,   2.3,  19. ,
        15. ,   3.1,   2.5,  14.7,   9.8,  12.2,  -1.5,  10.7,  13. ,
        26. ,  21. ,  14. ,  22. ,  25. ,   1.1,   1.6,  10.4,  27. ,
         1.9,   1. ,

In [88]:
df['ypa'].median()

6.7

In [89]:
df['ypa'].value_counts().head(20)

0.0    736
7.0    303
6.2    282
6.7    267
6.8    259
5.9    256
6.4    253
6.0    253
6.6    250
6.9    243
7.4    239
7.2    239
6.1    239
5.8    238
7.1    234
6.3    232
6.5    228
7.5    228
5.7    224
5.0    219
Name: ypa, dtype: int64

In [129]:
df['ypa'].isnull().values.any()

False

The median and mean were almost identical again at 6.8 and 6.7. The value counts were spread out almost evenly. The standard deviation was 4.22

### 2.7.13 TD per CMP Column<a id='2.7.13_TD_per_CMP_Column'></a>

In [90]:
df['td_per_cmp'].describe()

count    13185.000000
mean         0.070814
std          0.114830
min          0.000000
25%          0.000000
50%          0.053000
75%          0.100000
max          1.000000
Name: td_per_cmp, dtype: float64

In [91]:
df['td_per_cmp'].unique()

array([0.   , 0.125, 0.118, 0.043, 0.05 , 0.062, 0.095, 0.1  , 0.2  ,
       0.053, 0.25 , 0.042, 0.091, 1.   , 0.077, 0.111, 0.214, 0.074,
       0.067, 0.038, 0.133, 0.083, 0.048, 0.071, 0.333, 0.037, 0.158,
       0.143, 0.176, 0.08 , 0.04 , 0.045, 0.056, 0.12 , 0.231, 0.136,
       0.105, 0.059, 0.167, 0.087, 0.13 , 0.222, 0.115, 0.5  , 0.182,
       0.3  , 0.108, 0.15 , 0.16 , 0.103, 0.036, 0.065, 0.273, 0.097,
       0.148, 0.188, 0.031, 0.086, 0.286, 0.19 , 0.154, 0.211, 0.278,
       0.061, 0.069, 0.267, 0.034, 0.033, 0.032, 0.235, 0.107, 0.029,
       0.179, 0.308, 0.185, 0.174, 0.667, 0.088, 0.129, 0.375, 0.217,
       0.227, 0.057, 0.139, 0.051, 0.192, 0.03 , 0.094, 0.028, 0.023,
       0.208, 0.138, 0.054, 0.294, 0.027, 0.261, 0.156, 0.238, 0.026,
       0.4  , 0.161, 0.429, 0.385, 0.364, 0.207, 0.022, 0.073, 0.075,
       0.079, 0.121, 0.025, 0.194, 0.116, 0.081, 0.259, 0.263, 0.318,
       0.24 , 0.114, 0.047])

In [92]:
df['td_per_cmp'].median()

0.053

In [93]:
df['td_per_cmp'].value_counts().head(20)

0.000    4793
0.100     308
0.083     304
0.111     304
0.125     291
0.077     286
0.067     285
0.059     283
0.091     283
0.143     283
0.071     281
0.062     278
0.056     265
0.053     246
0.045     240
0.050     234
0.048     232
0.043     207
0.167     205
0.095     181
Name: td_per_cmp, dtype: int64

In [128]:
df['td_per_cmp'].isnull().values.any()

False

There are a few different unique values in this column which should help us to differentiate how well a QB preformed. Most of the value counts were 0.0 but many are spread out evenly as well.

### 2.7.14 TD per ATT Column<a id='2.7.14_TD_per_ATT_Column'></a>

In [94]:
df['td_per_att'].describe()

count    13088.000000
mean         0.047821
std          0.104252
min          0.000000
25%          0.000000
50%          0.031000
75%          0.061000
max          1.000000
Name: td_per_att, dtype: float64

In [95]:
df['td_per_att'].unique()

array([0.   , 0.08 , 0.065, 0.025, 0.024, 0.043, 0.054, 0.034, 0.077,
       0.059, 0.148, 0.026, 0.083, 0.061, 0.037, 1.   , 0.042, 0.062,
       0.136, 0.053, 0.05 , 0.022, 0.049, 0.03 , 0.333, 0.071, 0.111,
         nan, 0.087, 0.097, 0.086, 0.074, 0.033, 0.023, 0.068, 0.13 ,
       0.032, 0.091, 0.029, 0.04 , 0.027, 0.118, 0.16 , 0.167, 0.038,
       0.143, 0.1  , 0.028, 0.019, 0.176, 0.069, 0.045, 0.081, 0.103,
       0.067, 0.107, 0.094, 0.036, 0.057, 0.098, 0.047, 0.06 , 0.105,
       0.051, 0.016, 0.089, 0.088, 0.056, 0.073, 0.031, 0.114, 0.125,
       0.041, 0.048, 0.075, 0.02 , 0.214, 0.07 , 0.154, 0.138, 0.2  ,
       0.25 , 0.079, 0.188, 0.133, 0.095, 0.021, 0.161, 0.115, 0.044,
       0.182, 0.064, 0.12 , 0.085, 0.039, 0.052, 0.19 , 0.082, 0.129,
       0.018, 0.667, 0.102, 0.121, 0.217, 0.158, 0.15 , 0.093, 0.152,
       0.058, 0.222, 0.122, 0.035, 0.211, 0.174, 0.156, 0.286, 0.017,
       0.147, 0.055, 0.231, 0.24 , 0.5  , 0.108, 0.135, 0.185, 0.128,
       0.116, 0.172,

In [96]:
df['td_per_att'].median()

0.031

In [97]:
df['td_per_att'].value_counts().head(20)

0.000    4696
0.029     308
0.026     226
0.043     198
0.030     195
0.067     195
0.056     190
0.036     188
0.032     186
0.040     181
0.033     176
0.038     174
0.024     173
0.031     172
0.048     170
0.037     167
0.059     167
0.077     164
0.083     160
0.034     159
Name: td_per_att, dtype: int64

In [127]:
df['td_per_att'].isnull().values.any()

True

Similar to the td per cmp column we have a wide range of unique data points. The value counts are spread out as well.

### 2.7.15 Loss YDS Column<a id='2.7.15_Loss_YDS_Column'></a>

In [101]:
df['loss_yds'].describe()

count    13171.000000
mean        11.971225
std         11.970053
min          0.000000
25%          0.000000
50%          9.000000
75%         18.000000
max         91.000000
Name: loss_yds, dtype: float64

In [102]:
df['loss_yds'].unique()

array([11.,  0., 53., 12., 21., 15.,  8.,  7., 24., 23., 25., 14., 35.,
       26., 18.,  4.,  6., 16., 56., 31., 19.,  5., 29., 32.,  9., 17.,
       10., 13.,  2., 27.,  3., 22., 41.,  1., 20., 46., 47., 40., 37.,
       71., 42., 38., 45., 30., 34., 39., 33., 54., 28., 43., 50., 52.,
       55., 36., 49., 58., 66., 63., 60., 48., 57., 51., 67., 44., 59.,
       69., 62., 65., 68., 73., nan, 64., 70., 91., 72.])

In [103]:
df['loss_yds'].median()

9.0

In [104]:
df['loss_yds'].value_counts().head(20)

0.0     3501
8.0      575
7.0      550
9.0      485
6.0      437
10.0     437
5.0      376
14.0     370
15.0     351
16.0     349
11.0     345
12.0     344
17.0     325
13.0     323
18.0     291
19.0     247
3.0      238
21.0     230
20.0     228
22.0     228
Name: loss_yds, dtype: int64

In [126]:
df['loss_yds'].isnull().values.any()

True

We have 13,171 values. The data is missing a few. The mean and standard deviation are very similar at 11.971 and 11.970. Loss yards are from when a QB gets sacked and are considered negitice yardage.

### 2.7.16 YPC Column<a id='2.7.16_YPC_Column'></a>

In [105]:
df['ypc'].describe()

count    13185.000000
mean        11.145203
std          5.081139
min        -11.000000
25%          9.300000
50%         11.100000
75%         13.100000
max         81.000000
Name: ypc, dtype: float64

In [108]:
df['ypc'].unique()

array([  9.5,  12.2,  10.6,  13.4,  11.6,   9.8,  11.8,  13. ,  -8. ,
        11.4,   9.1,  13.5,  11.3,   9.9,   9.2,   8. ,   0. ,  12.4,
        24. ,  11. ,  10.9,  11.1,   7.1,  20.8,  10.5,   7. ,  15.4,
        33. ,  47. ,  13.2,   9.4,  16.1,  11.2,  12.3,  18.3,  10. ,
        12.6,   8.3,   4.5,  10.2,  12.5,   7.6,   5.5,  12. ,  11.5,
         6.5,  10.7,  14.3,  12.7,  14.2,  11.7,   5. ,   8.7,  10.1,
        18. ,  14.1,  19.8,   8.1,  12.8,  10.8,  12.1,  10.3,  13.1,
        16.7,   7.2,   7.7,   3. ,  15. ,   6.8,  10.4,   7.9,  14. ,
        18.8,   8.8,  13.6,   6.3,  14.6,  21. ,   8.5,  25.5,   8.6,
         9.3,   9. ,  16.9,  11.9,  15.6,  18.6,  17.1,  13.8,  13.7,
        16. ,   4. ,  16.6,  15.2,   8.9,   9.7,  16.2,  17.7,   7.5,
        17.8,  17.2,  13.3,   8.4,  14.8,  14.4,  15.3,   9.6,  15.5,
        14.9,  14.7,  17.9,  13.9,  12.9,  14.5,  16.5,  15.9,  15.1,
         6. ,  22.7,  30. ,  20.3,  -4. ,  27.5,  17.6,  18.5,  24.2,
        19.4,  16.3,

In [109]:
df['ypc'].median()

11.1

In [110]:
df['ypc'].value_counts().head(20)

0.0     736
11.0    227
10.0    225
11.1    221
11.5    203
10.8    198
11.2    197
10.4    195
10.6    195
12.2    195
12.0    193
11.6    186
10.5    183
11.4    183
11.3    182
11.8    179
10.2    176
10.9    176
9.8     174
10.7    171
Name: ypc, dtype: int64

In [125]:
df['ypc'].isnull().values.any()

False

There are many 0.0 values in this column. A wide range of unique values and a max of 81 yards per catch.

### 2.7.17 Home & Away Column<a id='2.7.17_Home_Away_Column'></a>

In [111]:
df['home_away'].describe()

count     13185
unique        2
top        home
freq       6628
Name: home_away, dtype: object

In [113]:
df['home_away'].unique()

array(['away', 'home'], dtype=object)

In [115]:
df['home_away'].value_counts().head(20)

home    6628
away    6557
Name: home_away, dtype: int64

In [124]:
df['home_away'].isnull().values.any()

False

There are 13,185 total games, (6,628+6,557). We will need to investigate why there are more home games than away if we have all the data for each game.

### 2.7.18 Year Column<a id='2.7.18_Year_Column'></a>

In [116]:
df['year'].describe()

count    13185.000000
mean      2005.898218
std          6.013394
min       1996.000000
25%       2001.000000
50%       2006.000000
75%       2011.000000
max       2016.000000
Name: year, dtype: float64

In [117]:
df['year'].unique()

array([1996, 1997, 1998, 1999, 2000, 2001, 2002, 2003, 2004, 2005, 2006,
       2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016],
      dtype=int64)

In [118]:
df['year'].value_counts().head(20)

2009    668
2003    663
1999    649
2004    648
2002    647
1998    642
2000    640
2007    639
2005    637
2008    626
2010    625
1997    624
1996    621
2006    620
2014    616
2011    614
2001    611
2016    610
2012    604
2015    592
Name: year, dtype: int64

In [123]:
df['year'].isnull().values.any()

False

Here we have more data on the year 2009 than 2015 by 76 games. Also why other years have different values if all 32 teams play 16 games a year, (for a total of 512 games per year). We should investigate this at a later stage in the exploration. 

### 2.8 Taking a Look at the Data as a Whole<a id='2.8_Taking_a_Look_at_the_Data_as_a_Whole'></a>

In [120]:
df.describe()

cmp           att        comp %           yds            td  \
count  13185.000000  13185.000000  13185.000000  13185.000000  13185.000000   
mean      16.126052     26.876071     58.110944    186.177722      1.118392   
std        8.771335     13.580485     21.333737    105.604116      1.119667   
min        0.000000      0.000000      0.000000    -11.000000      0.000000   
25%       11.000000     20.000000     51.200000    115.000000      0.000000   
50%       17.000000     29.000000     60.000000    197.000000      1.000000   
75%       22.000000     36.000000     67.600000    260.000000      2.000000   
max       58.000000     69.000000    500.000000    527.000000      7.000000   

                int          rate          long         sack   game_points  \
count  13185.000000  13185.000000  13185.000000  13185.00000  13185.000000   
mean       0.801062     80.157368     33.315738      1.86113     21.379446   
std        0.970170     32.157365     18.320263      1.70127     10.663774   
min        0.000000      0.000000    -11.000000      0.00000      0.000000   
25%        0.000000     58.500000     22.000000      0.00000     13.000000   
50%        1.000000     80.900000     31.000000      2.00000     21.000000   
75%        1.000000    102.000000     44.000000      3.00000     28.000000   
max        7.000000    158.300000     99.000000     12.00000     62.000000   

                ypa           ypc    td_per_cmp    td_per_att      loss_yds  \
count  13185.000000  13185.000000  13185.000000  13088.000000  13171.000000   
mean       6.884475     11.145203      0.070814      0.047821     11.971225   
std        4.225494      5.081139      0.114830      0.104252     11.970053   
min      -11.000000    -11.000000      0.000000      0.000000      0.000000   
25%        5.200000      9.300000      0.000000      0.000000      0.000000   
50%        6.700000     11.100000      0.053000      0.031000      9.000000   
75%        8.100000     13.100000      0.100000      0.061000     18.000000   
max       81.000000     81.000000      1.000000      1.000000     91.000000   

               year  
count  13185.000000  
mean    2005.898218  
std        6.013394  
min     1996.000000  
25%     2001.000000  
50%     2006.000000  
75%     2011.000000  
max     2016.000000

In [141]:
df.T.describe()

0      1      2      3                        4      \
count      18     18     18     18                       18   
unique     15     17     15     15                       18   
top         0      0      0      0  Kerry CollinsK. Collins   
freq        4      2      4      4                        1   

                     5      6      7      8      9      ...  13178 13179  \
count                   18     18     18     18     18  ...     18    18   
unique                  18     16     17      9     17  ...      9     6   
top     Jeff BlakeJ. Blake      0     13      0      1  ...      0     0   
freq                     1      3      2      6      2  ...      6    12   

        13180  13181  13182  13183  13184  13185  13186  \
count      18     18     18     18     18     18     18   
unique     17     12     17     16     16     17     15   
top         1      0      0      2      2      1      0   
freq        2      6      2      2      2      2      4   

                                13187  
count                              18  
unique                             18  
top     Colin KaepernickC. Kaepernick  
freq                                1  

[4 rows x 13185 columns]

In [141]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 13185 entries, 0 to 13187
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   qb           13185 non-null  object 
 1   cmp          13185 non-null  int64  
 2   att          13185 non-null  int64  
 3   comp %       13185 non-null  float64
 4   yds          13185 non-null  float64
 5   td           13185 non-null  int64  
 6   int          13185 non-null  int32  
 7   rate         13185 non-null  float64
 8   long         13185 non-null  int32  
 9   sack         13185 non-null  float64
 10  game_points  13185 non-null  int64  
 11  ypa          13185 non-null  float64
 12  ypc          13185 non-null  float64
 13  td_per_cmp   13185 non-null  float64
 14  td_per_att   13088 non-null  float64
 15  loss_yds     13171 non-null  float64
 16  home_away    13185 non-null  object 
 17  year         13185 non-null  int64  
dtypes: float64(9), int32(2), int64(5), object(2)
m

Here we can see that we have some work to do to clean the data. td_per_att and loss_yds are missing some values. We have floats and integers as values.

### 2.9 Summary<a id='2.9_Summary'></a>

The data is very informative but we still have some work to do for us to make a new QB rating system. In this part of the data exploration we added rows, switched around the columns, cleaned some of the data and found key features of each columns like the standard deviation. This is a good foundation for the next chapter in the data exploration.